### PR TITLE
[staging] {cc,binutils}-wrapper: fixes for PIE hardening

### PR DESCRIPTION
**Pull request by r-burns**
Fixes a number of packages which were broken on musl (where PIE hardening is currently enabled). For example:
* pkgsMusl.python3
* pkgsMusl.bulletml
* pkgsMusl.proot
* pkgsMusl.libfsm
* pkgsMusl.libiscsi
* pkgsMusl.nsjail
* pkgsMusl.pv

###### Motivation for this change

Fixes https://github.com/NixOS/nixpkgs/issues/124476

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-burns** [comment]
cc @TredwellGit since it looks like you're running into some of these issues in https://github.com/NixOS/nixpkgs/pull/104091

File: pkgs/build-support/cc-wrapper/add-hardening.sh [hunk 1]
```diff
@@ -34,6 +34,16 @@ if (( "${NIX_DEBUG:-0}" >= 1 )); then
   fi
 fi
 
+# Ensure PIE flags occur before PIC, if applicable
+if [ "${hardeningEnableMap[pie]-}" = 1 ]; then
+  if (( "${NIX_DEBUG:-0}" >= 1 )); then echo HARDENING: enabling CFlags -fPIE >&2; fi
+  hardeningCFlags+=('-fPIE')
```
Comment (veprbl) [inline pkgs/build-support/cc-wrapper/add-hardening.sh:40]: Can we instead do
`hardeningCFlags =( '-fPIE' "${hardeningCFlags[@]}" )`

Comment (r-burns) [reply pkgs/build-support/cc-wrapper/add-hardening.sh:40]: Great idea, that is SO much cleaner than what I have. Gonna run the rebuild gauntlet on this and get back to you.

Comment (r-burns) [reply pkgs/build-support/cc-wrapper/add-hardening.sh]: Works great!

**Review by veprbl — APPROVED** [review]

**Pull request merged**